File: compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaSynthesizer.kt
/*
 * Copyright 2026 The Google AI Edge Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ai.edge.examples.text_to_speech

import android.content.Context
import com.google.ai.edge.litert.Accelerator
import com.google.ai.edge.litert.CompiledModel
import java.io.Closeable
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.Random
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.sin

/**
 * Matcha-TTS on-device. The three heavy graphs run on the LiteRT CompiledModel GPU (ML Drift);
 * everything stochastic/sequential runs host-side here (no FFT anywhere — Matcha's HiFi-GAN
 * vocoder is a time-domain GAN, which is what lets the whole synthesis path stay on the GPU).
 *
 * ```
 * phoneme ids --(host: embed + pad)--> text_encoder(GPU) --> mu, logw
 *             --(host: durations + length-regulator)--> mu_y[1,80,T]
 *             --(host: Euler ODE loop, N steps)-->
 *                 decoder(GPU) x N  (x_t, mu_y, sin_emb(t), mask) --> v
 *             --(host: denormalize)--> mel --> vocoder(GPU) --> waveform[T*256]
 * ```
 *
 * Fixed shapes ([MAX_TEXT] phonemes, [MAX_MEL] frames); a runtime float mask makes the padded
 * positions a no-op (additive attention bias), so one compiled graph handles any length.
 * Graphs are loaded from filesDir (push via scripts/install_to_device.sh).
 */
class MatchaSynthesizer(private val context: Context) : Closeable {

    /** One synthesis result: float PCM plus timing metadata for the demo UI. */
    data class Result(val audio: FloatArray, val frames: Int, val steps: Int, val ms: Long)

    private val embeddingTable: FloatArray =
        readFloats(context.assets.open("emb.bin").readBytes()) // [178 * N_CHANNELS]

    private val textEncoder = loadModel(TEXT_ENCODER_MODEL)

    // Decoder on CPU: the diffusers transformer blocks are mis-fused on the Mali ML Drift
    // delegate (residual collapses to corr 0.006 in the full graph, though the SAME
    // transformer is corr 0.98 as a standalone graph -> a fusion bug, not the ops). On CPU
    // the decoder is exact; RTF stays realtime (~0.93). Text encoder + vocoder run on GPU.
    private val decoder = loadModel(DECODER_MODEL, Accelerator.CPU)
    private val vocoder = loadModel(VOCODER_MODEL)

    private val textEncoderInputs = textEncoder.createInputBuffers()
    private val textEncoderOutputs = textEncoder.createOutputBuffers()
    private val decoderInputs = decoder.createInputBuffers()
    private val decoderOutputs = decoder.createOutputBuffers()
    private val vocoderInputs = vocoder.createInputBuffers()
    private val vocoderOutputs = vocoder.createOutputBuffers()

    private fun loadModel(name: String, accelerator: Accelerator = Accelerator.GPU): CompiledModel {
        val file = File(context.filesDir, name)
        check(file.exists()) {
            "Model not found: $name. Push it first:\n  scripts/install_to_device.sh"
        }
        return CompiledModel.create(file.absolutePath, CompiledModel.Options(accelerator), null)
    }

    /**
     * Synthesizes speech from matcha symbol ids (NOT yet interspersed with blanks).
     *
     * @param phonemeIds matcha symbol ids from [MatchaG2P.phonemize].
     * @param nSteps number of Euler ODE steps (more = higher quality, slower).
     * @param seed optional fixed noise seed for reproducible output.
     */
    fun synthesize(phonemeIds: IntArray, nSteps: Int = DEFAULT_STEPS, seed: Long? = null): Result {
        val startNanos = System.nanoTime()

        // Intersperse blanks (id 0), pad to MAX_TEXT, build the text mask.
        val textLength = minOf(phonemeIds.size * 2 + 1, MAX_TEXT)
        val ids = IntArray(MAX_TEXT)
        var writeIndex = 1
        for (id in phonemeIds) {
            if (writeIndex >= MAX_TEXT) break
            ids[writeIndex] = id
            writeIndex += 2
        }
        val textMask = FloatArray(MAX_TEXT) { if (it < textLength) 1f else 0f }

        // Host phoneme-embedding lookup -> [1, MAX_TEXT, N_CHANNELS].
        val embedded = FloatArray(MAX_TEXT * N_CHANNELS)
        for (t in 0 until MAX_TEXT) {
            System.arraycopy(
                embeddingTable, ids[t] * N_CHANNELS, embedded, t * N_CHANNELS, N_CHANNELS)
        }

        // Text encoder (GPU) -> mu[1,80,T], logw[1,1,T].
        textEncoderInputs[0].writeFloat(embedded)
        textEncoderInputs[1].writeFloat(textMask)
        textEncoder.run(textEncoderInputs, textEncoderOutputs)
        val mu = textEncoderOutputs[0].readFloat() // [N_FEATS * MAX_TEXT], channel-major
        val logDurations = textEncoderOutputs[1].readFloat() // [MAX_TEXT]

        // Durations -> cumulative -> length regulator (mu_y).
        val durations = FloatArray(MAX_TEXT) {
            ceil(exp(logDurations[it]) * textMask[it]) * LENGTH_SCALE
        }
        val cumulative = FloatArray(MAX_TEXT)
        var total = 0f
        for (i in 0 until MAX_TEXT) {
            total += durations[i]
            cumulative[i] = total
        }
        val melLength = minOf(maxOf(total.toInt(), 1), MAX_MEL)

        val muY = FloatArray(N_FEATS * MAX_MEL)
        var phonemeIndex = 0
        for (frame in 0 until melLength) {
            while (phonemeIndex < MAX_TEXT - 1 && cumulative[phonemeIndex] <= frame) {
                phonemeIndex++
            }
            for (c in 0 until N_FEATS) {
                muY[c * MAX_MEL + frame] = mu[c * MAX_TEXT + phonemeIndex]
            }
        }
        val melMask = FloatArray(MAX_MEL) { if (it < melLength) 1f else 0f }

        // Euler ODE: x_{k+1} = x_k + dt * decoder(x_k, mu_y, sin_emb(t), mask).
        val random = if (seed != null) Random(seed) else Random()
        val x = FloatArray(N_FEATS * MAX_MEL)
        for (c in 0 until N_FEATS) {
            for (frame in 0 until melLength) {
                x[c * MAX_MEL + frame] = random.nextGaussian().toFloat()
            }
        }
        val dt = 1f / nSteps
        var time = 0f
        repeat(nSteps) {
            decoderInputs[0].writeFloat(x)
            decoderInputs[1].writeFloat(muY)
            decoderInputs[2].writeFloat(sinusoidalPositionEmbedding(time))
            decoderInputs[3].writeFloat(melMask)
            decoder.run(decoderInputs, decoderOutputs)
            val velocity = decoderOutputs[0].readFloat()
            for (i in x.indices) {
                x[i] += dt * velocity[i]
            }
            time += dt
        }

        // Denormalize + zero pad -> mel -> vocoder (GPU) -> waveform.
        val mel = FloatArray(N_FEATS * MAX_MEL)
        for (c in 0 until N_FEATS) {
            for (frame in 0 until melLength) {
                val i = c * MAX_MEL + frame
                mel[i] = x[i] * MEL_STD + MEL_MEAN
            }
        }
        vocoderInputs[0].writeFloat(mel)
        vocoder.run(vocoderInputs, vocoderOutputs)
        val waveform = vocoderOutputs[0].readFloat() // [MAX_MEL * HOP]
        val sampleCount = melLength * HOP
        val audio = FloatArray(sampleCount) { waveform[it].coerceIn(-1f, 1f) }

        return Result(audio, melLength, nSteps, (System.nanoTime() - startNanos) / 1_000_000)
    }

    /** Matcha SinusoidalPosEmb (weight-free): t -> [TIME_DIM] = [sin(...) | cos(...)]. */
    private fun sinusoidalPositionEmbedding(t: Float, scale: Float = 1000f): FloatArray {
        val half = TIME_DIM / 2
        val embedding = FloatArray(TIME_DIM)
        val logScale = -ln(10000.0) / (half - 1)
        for (i in 0 until half) {
            val angle = scale * t * exp(i * logScale).toFloat()
            embedding[i] = sin(angle)
            embedding[half + i] = cos(angle)
        }
        return embedding
    }

    private fun readFloats(bytes: ByteArray): FloatArray {
        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
        return FloatArray(bytes.size / 4) { buffer.float }
    }

    override fun close() {
        textEncoderInputs.forEach { it.close() }
        textEncoderOutputs.forEach { it.close() }
        decoderInputs.forEach { it.close() }
        decoderOutputs.forEach { it.close() }
        vocoderInputs.forEach { it.close() }
        vocoderOutputs.forEach { it.close() }
        textEncoder.close()
        decoder.close()
        vocoder.close()
    }

    companion object {
        const val MAX_TEXT = 256
        const val MAX_MEL = 512
        const val N_FEATS = 80
        const val N_CHANNELS = 192
        const val HOP = 256
        const val SAMPLE_RATE = 22050
        const val LENGTH_SCALE = 0.95f
        const val MEL_MEAN = -5.536622f
        const val MEL_STD = 2.116101f
        const val TIME_DIM = 160
        const val DEFAULT_STEPS = 10
        const val MAX_SAMPLES = MAX_MEL * HOP

        private const val TEXT_ENCODER_MODEL = "matcha_textenc_fp16.tflite"
        private const val DECODER_MODEL = "matcha_decoder_fp16.tflite"
        private const val VOCODER_MODEL = "matcha_vocoder_fp16.tflite"
    }
}

File: samples/litert/text_to_speech/ios/Sources/MatchaRuntime/MatchaSynthesizer.swift
// Copyright 2026 The Google AI Edge Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation
import LiteRT

// The package exposes MatchaCore; the app compiles these sources into one module.
#if canImport(MatchaCore)
  import MatchaCore
#endif

/// Invalid synthesis configuration or nonfinite intermediate tensors.
public enum MatchaSynthesizerError: Error, LocalizedError {
  case unsupportedSignature(String)
  case invalidStepCount(Int)
  case nonfiniteTensor(stage: String, index: Int)

  public var errorDescription: String? {
    switch self {
    case .unsupportedSignature(let reason): return reason
    case .invalidStepCount(let count): return "The step count must be positive (received \(count))."
    case .nonfiniteTensor(let stage, let index):
      return "Synthesis stopped: \(stage) contains a non-finite value at index \(index)."
    }
  }
}

/// Independent accelerator configurations for the three acoustic graphs.
public struct MatchaPlacement: Codable {
  public var textEncoder: MatchaGraphConfiguration
  public var decoder: MatchaGraphConfiguration
  public var vocoder: MatchaGraphConfiguration

  /// Selects the accelerator and precision independently for each acoustic graph.
  public init(
    textEncoder: MatchaGraphConfiguration = .cpu, decoder: MatchaGraphConfiguration = .cpu,
    vocoder: MatchaGraphConfiguration = .gpu
  ) {
    self.textEncoder = textEncoder
    self.decoder = decoder
    self.vocoder = vocoder
  }

  public static let cpuOnly = MatchaPlacement(vocoder: .cpu)
}

/// Generated audio, acoustic controls, optional tensors, and graph measurements.
public struct MatchaSynthesis {
  public let audio: [Float]
  public let textLength: Int
  public let paddedIDs: [Int32]
  public let durations: [Float]
  public let melLength: Int
  public let steps: Int
  public let dt: Float
  public let times: [Float]
  public let tensors: [String: [Float]]
  public let graphTimings: [String: GraphTiming]
  public let graphStatuses: [String: GraphBackendStatus]
}

/// Owns reused graph buffers. Calls to synthesize are serialized for their full duration.
public final class MatchaSynthesizer {
  private let textEncoder: LiteRTGraph
  private let decoder: LiteRTGraph
  private let vocoder: LiteRTGraph
  private let embedding: [Float]
  private let muIndex: Int
  private let logwIndex: Int
  private let lock = NSLock()
  public let textEncoderOutputNames: [String]
  public let textEncoderOutputShapes: [[Int]]
  public let muOutputName: String
  public let logwOutputName: String

  /// Pass the same environment that the G2P runner uses.
  public init(assets: URL, environment: Environment, placement: MatchaPlacement = MatchaPlacement())
    throws
  {
    embedding = try MatchaMath.readLittleEndianFloats(
      Data(contentsOf: assets.appendingPathComponent("emb.bin")))
    let encoderGraph = try LiteRTGraph(
      modelPath: assets.appendingPathComponent("matcha_textenc_fp16.tflite").path,
      environment: environment, configuration: placement.textEncoder)
    let decoderGraph = try LiteRTGraph(
      modelPath: assets.appendingPathComponent("matcha_decoder_fp16.tflite").path,
      environment: environment, configuration: placement.decoder)
    let vocoderGraph = try LiteRTGraph(
      modelPath: assets.appendingPathComponent("matcha_vocoder_fp16.tflite").path,
      environment: environment, configuration: placement.vocoder)
    let muIndices = encoderGraph.outputShapes.indices.filter {
      encoderGraph.outputShapes[$0] == [1, 80, 256]
    }
    let logwIndices = encoderGraph.outputShapes.indices.filter {
      encoderGraph.outputShapes[$0] == [1, 1, 256]
    }
    guard muIndices.count == 1 && logwIndices.count == 1 && encoderGraph.outputShapes.count == 2,
      encoderGraph.inputShapes == [[1, 256, 192], [1, 1, 256]],
      decoderGraph.inputShapes == [[1, 80, 512], [1, 80, 512], [1, 160], [1, 1, 512]],
      decoderGraph.outputShapes == [[1, 80, 512]],
      vocoderGraph.inputShapes == [[1, 80, 512]],
      vocoderGraph.outputShapes == [[1, 1, 131072]]
    else {
      throw MatchaSynthesizerError.unsupportedSignature("Unexpected acoustic graph shapes")
    }
    textEncoder = encoderGraph
    decoder = decoderGraph
    vocoder = vocoderGraph
    muIndex = muIndices[0]
    logwIndex = logwIndices[0]
    textEncoderOutputNames = encoderGraph.outputNames
    textEncoderOutputShapes = encoderGraph.outputShapes
    muOutputName = encoderGraph.outputNames[muIndices[0]]
    logwOutputName = encoderGraph.outputNames[logwIndices[0]]
  }

  /// Executes text encoding, length regulation, Euler integration, and vocoding.
  public func synthesize(
    ids: [Int32], steps: Int = MatchaMath.defaultSteps, seed: UInt64 = 0,
    noise: [Float]? = nil, captureTensors: Bool = false
  ) throws -> MatchaSynthesis {
    lock.lock()
    defer { lock.unlock() }
    guard steps > 0 else {
      throw MatchaSynthesizerError.invalidStepCount(steps)
    }
    var tensors = [String: [Float]]()
    var timings = [String: GraphTiming]()
    let text = try MatchaMath.prepareText(ids: ids, embedding: embedding)
    let encoded = try textEncoder.run([text.embedded, text.mask])
    timings["text_encoder"] = textEncoder.lastTiming
    let mu = encoded[muIndex]
    let logw = encoded[logwIndex]
    try Self.requireFinite(mu, stage: "text encoder mu")
    try Self.requireFinite(logw, stage: "text encoder logw")
    let regulated = try MatchaMath.regulate(mu: mu, logw: logw, textMask: text.mask)
    try Self.requireFinite(regulated.durations, stage: "durations")
    try Self.requireFinite(regulated.cumulative, stage: "cumulative durations")
    var random = GaussianNoise(seed: seed)
    var x =
      try noise.map { try MatchaMath.maskNoise($0, melLength: regulated.melLength) }
      ?? random.tensor(melLength: regulated.melLength)
    try Self.requireFinite(x, stage: "initial noise")
    if captureTensors {
      tensors["embedded"] = text.embedded
      tensors["text_mask"] = text.mask
      tensors["mu"] = mu
      tensors["logw"] = logw
      tensors["durations"] = regulated.durations
      tensors["cumulative"] = regulated.cumulative
      tensors["mu_y"] = regulated.muY
      tensors["mel_mask"] = regulated.mask
      tensors["x0"] = x
    }
    let dt: Float = 1 / Float(steps)
    var time: Float = 0
    var times = [Float]()
    for step in 0..<steps {
      let prefix = String(format: "step_%02d", step)
      let timeEmbedding = MatchaMath.timeEmbedding(time)
      let velocity = try decoder.run([x, regulated.muY, timeEmbedding, regulated.mask])[0]
      timings[prefix] = decoder.lastTiming
      try Self.requireFinite(velocity, stage: "decoder velocity at step \(step)")
      try MatchaMath.eulerUpdate(x: &x, velocity: velocity, dt: dt)
      try Self.requireFinite(x, stage: "Euler output at step \(step)")
      if captureTensors {
        tensors[prefix + "_time"] = timeEmbedding
        tensors[prefix + "_velocity"] = velocity
        tensors[prefix + "_x"] = x
      }
      times.append(time)
      time = time + dt
    }
    let mel = try MatchaMath.denormalize(x, melLength: regulated.melLength)
    try Self.requireFinite(mel, stage: "mel")
    let fullWaveform = try vocoder.run([mel])[0]
    try Self.requireFinite(fullWaveform, stage: "vocoder waveform")
    timings["vocoder"] = vocoder.lastTiming
    let waveform = try MatchaMath.crop(fullWaveform, melLength: regulated.melLength)
    let audio = MatchaMath.clip(waveform)
    if captureTensors {
      tensors["mel"] = mel
      tensors["vocoder_full"] = fullWaveform
      tensors["waveform"] = waveform
      tensors["audio"] = audio
    }
    return MatchaSynthesis(
      audio: audio, textLength: text.textLength, paddedIDs: text.paddedIDs,
      durations: regulated.durations, melLength: regulated.melLength,
      steps: steps, dt: dt, times: times, tensors: tensors, graphTimings: timings,
      graphStatuses: [
        "text_encoder": textEncoder.backendStatus,
        "decoder": decoder.backendStatus, "vocoder": vocoder.backendStatus,
      ])
  }

  /// Rejects invalid values before they can affect another synthesis stage.
  private static func requireFinite(_ values: [Float], stage: String) throws {
    if let index = values.firstIndex(where: { !$0.isFinite }) {
      throw MatchaSynthesizerError.nonfiniteTensor(stage: stage, index: index)
    }
  }
}
